iree/compiler/Dialect/Flow/Transforms/InterchangeGenericOp.cpp

```diff
@@ -0,0 +1,78 @@
+// Copyright 2021 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+//===--------------- InterchangeGenericOp.cpp -----------------------------===//
+//
+// Interchange loops in generic ops to force the reduction loops to be the most
+// inner loops.
+//
+//===----------------------------------------------------------------------===//
+
+#include "iree/compiler/Dialect/Flow/Transforms/PassDetail.h"
+#include "iree/compiler/Dialect/Flow/Transforms/Passes.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Linalg/IR/LinalgOps.h"
+#include "mlir/Dialect/Linalg/Transforms/Transforms.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Flow {
+
+namespace {
+
+struct GenericOpInterchangePattern
+    : public OpRewritePattern<linalg::GenericOp> {
+  using OpRewritePattern<linalg::GenericOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(linalg::GenericOp genericOp,
+                                PatternRewriter &rewriter) const override {
+    SmallVector<unsigned> interchange;
+    bool needInterchange = false;
+    unsigned numParallelLoop = genericOp.getNumParallelLoops();
+    if (numParallelLoop == 0) return failure();
+    for (auto iter : llvm::enumerate(genericOp.iterator_types())) {
+      if (isParallelIterator(iter.value())) {
```

Comment:
Maybe too aggressive. The Flow only distributed upto 3 loops. Maybe just switch upto 3 dimensions?

Reply:
Do you expect it to be beneficial to restrict it to 3 dimensions? I thought we would always want to move the reduction to the inner loop no matter what, that what the front end has been trying to do. 
I assume we will want to distribute more than 3 dimensions in the future?

Reply:
I personally would like it if we keep it more general than 3 dimensions :)

Reply:
I am not convinced that there is a use case to distribute more than 3 dimensions. So my view there is more of a holding pattern. Moving "reduction" to inner most loop might be fine. Just seeing if we have any indicators to aid this decision. I am fine with moving it to the innermost.

Reply:
Sounds good, I'll keep it like that for now.